### ISSUE TITLE
Conv2D CPU outputs inconsistent with PyTorch baseline under identical weights/inputs

### ISSUE THREAD

**Issue by tingPetty** [tensorflow/tensorflow#105829]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.19

### Custom code

Yes

### OS platform and distribution

Windows 11

### Python version

3.10.18

### Current behavior?

When running Conv2D on CPU with identical weights, biases, and inputs, TensorFlow Conv2D produces outputs that significantly differ from PyTorch's Conv2D, even though both should compute the same operation. This occurs on multiple test cases.

Example:
- test474: max absolute difference ≈ 0.0005722

### Standalone code to reproduce the issue

```shell
import numpy as np
import torch
import tensorflow as tf


def run_single_test(name, input_tensor, weight, bias, stride, padding, dilation):
    print(f"\n===== {name} =====")

    # ===== PyTorch Conv2D 作为 baseline =====
    pt_conv = torch.nn.Conv2d(
        in_channels=weight.shape[1],
        out_channels=weight.shape[0],
        kernel_size=weight.shape[2:],
        stride=stride,
        padding=padding,
        dilation=dilation,
        bias=True,
    )

    # 覆盖参数
    with torch.no_grad():
        pt_conv.weight = torch.nn.Parameter(weight)
        pt_conv.bias = torch.nn.Parameter(bias)
        pt_out = pt_conv(input_tensor).cpu().numpy()

    # ===== TensorFlow Conv2D =====
    # PyTorch (O, I, KH, KW) -> TF (KH, KW, I, O)
    tf_weight = tf.convert_to_tensor(weight.permute(2, 3, 1, 0).numpy())
    tf_bias = tf.convert_to_tensor(bias.numpy())

    # TF padding
    padding_type = "VALID" if padding == (0, 0) else "SAME"

    tf_conv = tf.keras.layers.Conv2D(
        filters=weight.shape[0],
        kernel_size=weight.shape[2:],
        strides=stride,
        dilation_rate=dilation,
        padding=padding_type,
        use_bias=True,
        kernel_initializer=tf.constant_initializer(tf_weight.numpy()),
        bias_initializer=tf.constant_initializer(tf_bias.numpy()),
    )

    tf_input = tf.convert_to_tensor(input_tensor.numpy().transpose(0, 2, 3, 1))  # NCHW -> NHWC
    tf_out = tf_conv(tf_input).numpy().transpose(0, 3, 1, 2)  # NHWC -> NCHW

    # ===== 比较结果 =====
    try:
        np.testing.assert_allclose(pt_out, tf_out, rtol=1e-5, atol=1e-5)
        print(f"[SUCCESS] PyTorch vs TF matched for {name}")
    except AssertionError as e:
        print(f"[FAILED] PyTorch vs TF differ for {name}: {str(e)}")


# ====== 运行你的三个测试 ======

run_single_test(
    "test406",
    torch.randn([1, 3, 224, 224], dtype=torch.float32),
    torch.randn([64, 3, 11, 11], dtype=torch.float32),
    torch.randn([64], dtype=torch.float32),
    stride=(4, 4),
    padding=(2, 2),
    dilation=(1, 1),
)

run_single_test(
    "test474",
    torch.randn([1, 384, 13, 13], dtype=torch.float32),
    torch.randn([256, 384, 3, 3], dtype=torch.float32),
    torch.randn([256], dtype=torch.float32),
    stride=(1, 1),
    padding=(1, 1),
    dilation=(1, 1),
)

run_single_test(
    "test1269",
    torch.randn([1, 20, 10, 10], dtype=torch.float32),
    torch.randn([20, 20, 5, 5], dtype=torch.float32),
    torch.randn([20], dtype=torch.float32),
    stride=(1, 1),
    padding=(0, 0),
    dilation=(1, 1),
)
```

### Relevant log output

```shell
2025-12-08 13:46:25.384299: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-12-08 13:46:26.183422: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.

===== test406 =====
2025-12-08 13:46:28.231164: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE3 SSE4.1 SSE4.2 AVX AVX2 AVX512F AVX512_VNNI AVX512_BF16 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
[FAILED] PyTorch vs TF differ for test406: 
Not equal to tolerance rtol=1e-05, atol=1e-05

(shapes (1, 64, 55, 55), (1, 64, 56, 56) mismatch)
 x: array([[[[ 2.985665e+01, -3.691774e+01, -1.824992e+01, ...,
           1.306543e+00, -2.764326e+01, -1.711653e+00],
         [-3.449988e+00, -2.018358e+00,  2.474983e+01, ...,...
 y: array([[[[-13.726848,  -2.397375, -11.96731 , ...,  35.3033  ,
           -3.149915, -16.397116],
         [ -1.42489 ,  15.709376, -14.36447 , ..., -13.128145,...

===== test474 =====
[FAILED] PyTorch vs TF differ for test474: 
Not equal to tolerance rtol=1e-05, atol=1e-05

Mismatched elements: 1035 / 43264 (2.39%)
Max absolute difference: 0.0005722
Max relative difference: 0.02086491
 x: array([[[[  19.62614 ,  -13.875781,   73.90697 , ...,  -34.175003,
             5.878419,  -10.207954],
         [  64.93527 ,  -44.894314,    4.439484, ...,   17.981272,...
 y: array([[[[  19.626148,  -13.875776,   73.90699 , ...,  -34.17494 ,
             5.878421,  -10.207946],
         [  64.93523 ,  -44.894226,    4.439514, ...,   17.981277,...

===== test1269 =====
[FAILED] PyTorch vs TF differ for test1269: 
Not equal to tolerance rtol=1e-05, atol=1e-05

Mismatched elements: 3 / 720 (0.417%)
Max absolute difference: 5.340576e-05
Max relative difference: 0.0002654
 x: array([[[[ 2.648679e+00,  2.291362e+01,  3.526057e+00,  8.759871e+00,
           2.791437e+01,  4.824985e+00],
         [ 3.135897e+01, -2.411813e+01, -3.018489e+01,  2.356510e+00,...
 y: array([[[[ 2.648689e+00,  2.291362e+01,  3.526056e+00,  8.759874e+00,
           2.791438e+01,  4.824987e+00],
         [ 3.135896e+01, -2.411813e+01, -3.018491e+01,  2.356488e+00,...
```

**Comment by akhilgoe** [comment]
Hi @tingPetty, I believe `padding="SAME"` in TF does not conform with the padding used in the corresponding PyTorch case (for atleast test406). I modified your test to explicitly add a padding layer before the conv layer and set the conv padding attribute to `"VALID"`. With this change, all the three tests pass for me with a `1e-4` tolerance.

**Comment by nileshau-afk** [comment]
Hi @tingPetty, 

Thank you for your patience, we have reproduced the issue [gist](https://colab.sandbox.google.com/gist/nileshau-afk/d3e070d8e63f9493ab4bc33353ed07be/105829.ipynb). As a workaround try to manually pad the TensorFlow input using tf.pad to exactly match PyTorch's padding, and then run Conv2D with "VALID" padding and disable the oneDNN fast-math routines by setting an environment variable at the top of the script before importing TensorFlow.
```
# Manually pad 2 pixels on all spatial borders (top/bottom/left/right)
tf_input = tf.pad(tf_input, [[0, 0], [2, 2], [2, 2], [0, 0]])
tf_conv = tf.keras.layers.Conv2D(..., padding="VALID")
```
```
import os
os.environ["TF_ENABLE_ONEDNN_OPTS"] = "0"
import tensorflow as tf
```
If the numerical differences persist after correcting the padding, try running your script with `TF_ENABLE_ONEDNN_OPTS=0` set in your environment to see if the results become more consistent. Could you please try this and let us know if it works for you?
